src/audio/igo_nr/igo_nr.c

```diff
@@ -0,0 +1,804 @@
+	if (!cd->model_handler) {
+		comp_cl_err(&comp_igo_nr, "igo_nr_new(): comp_data_blob_handler_new() failed.");
+		goto err;
+	}
+
+	/* Get configuration data */
+	ret = comp_init_data_blob(cd->model_handler, bs, ipc_igo_nr->data);
+	if (ret < 0) {
+		comp_cl_err(&comp_igo_nr, "igo_nr_new(): comp_init_data_blob() failed.");
+		goto err;
+	}
+	dev->state = COMP_STATE_READY;
+
+	comp_cl_info(&comp_igo_nr, "igo_nr created");
+
+	return dev;
+
+err:
+	rfree(dev);
+	rfree(cd);
+	return NULL;
+}
+
+static void igo_nr_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_info(dev, "igo_nr_free()");
+
+	comp_data_blob_handler_free(cd->model_handler);
+
+	rfree(cd->p_handle);
+	rfree(cd);
+	rfree(dev);
+}
+
+static int32_t igo_nr_verify_params(struct comp_dev *dev,
+				    struct sof_ipc_stream_params *params)
+{
+	int32_t ret;
+
+	comp_dbg(dev, "igo_nr_verify_params()");
+
+	/* update downstream (playback) or upstream (capture) buffer parameters
+	 */
+	ret = comp_verify_params(dev, BUFF_PARAMS_RATE, params);
+	if (ret < 0) {
+		comp_err(dev, "igo_nr_verify_params(): comp_verify_params() failed.");
+		return ret;
+	}
+
+	return 0;
+}
+
+/* set component audio stream parameters */
+static int32_t igo_nr_params(struct comp_dev *dev,
+			     struct sof_ipc_stream_params *pcm_params)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sinkb;
+	struct comp_buffer *sourceb;
+	int32_t err;
+
+	comp_info(dev, "igo_nr_params()");
+
+	err = igo_nr_verify_params(dev, pcm_params);
+	if (err < 0) {
+		comp_err(dev, "igo_nr_params(): pcm params verification failed.");
+		return -EINVAL;
+	}
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
+				  sink_list);
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer,
+				source_list);
+
+	/* set source/sink_frames/rate */
+	cd->source_rate = sourceb->stream.rate;
+	cd->sink_rate = sinkb->stream.rate;
+	if (!cd->sink_rate) {
+		comp_err(dev, "igo_nr_params(), zero sink rate");
+		return -EINVAL;
+	}
+
+	cd->sink_frames = dev->frames;
+	cd->source_frames = ceil_divide(dev->frames * cd->source_rate,
+					cd->sink_rate);
+
+	/* Use empirical add +10 for target frames number to avoid xruns and
+	 * distorted audio in beginning of streaming. It's slightly more than
+	 * min. needed and does not increase much peak load and buffer memory
+	 * consumption. The copy() function will find the limits and process
+	 * less frames after the buffers levels stabilize.
+	 */
+	cd->source_frames_max = cd->source_frames + 10;
+	cd->sink_frames_max = cd->sink_frames + 10;
+	cd->frames = MAX(cd->source_frames_max, cd->sink_frames_max);
+
+	comp_dbg(dev, "igo_nr_params(), source_rate=%u, sink_rate=%u, source_frames_max=%d, sink_frames_max=%d",
+		 cd->source_rate, cd->sink_rate,
+		 cd->source_frames_max, cd->sink_frames_max);
+
+	/* The igo_nr supports sample rate 16000 only. */
+	switch (sourceb->stream.rate) {
```

Comment:
do ou plan to support other formats in the future? if not, why not return -EINVAL is stream.rate != 16000 here?

Reply:
Yes. We'd like to reserve the possibility of implementing SRC inside our library. However this is not prioritized now because SOF FW seems to support SRC component already.